examples/pipelines/commoncrawl/components/download_commoncrawl_segments/src/main.py

```diff
@@ -0,0 +1,81 @@
+import logging
+import typing as t
+
+import io
+import boto3
+
+import dask.bag as db
+import dask.dataframe as dd
+from dask.delayed import delayed
+
+import pandas as pd
+
+import requests
+from warcio.archiveiterator import ArchiveIterator
+
+from fondant.component import DaskTransformComponent
+
+logger = logging.getLogger(__name__)
+
+BASE_URL = "https://data.commoncrawl.org/"
+
+
+def get_records_from_warc_file(
+    warc_file: str, n_records_to_download: t.Optional[int] = None
+) -> t.List:
+    """Downloads a WARC file and extracts the webpages.
+    Args:
+        warc_file: The path to the WARC file.
+        n_records_to_download: The number of webpages to download from the WARC file.
+    Returns:
+        A list of webpages.
+    """
+    logger.info(f"Processing WARC file from segment path: {warc_file}...")
+    records = []
+    counter = 0
+    response = requests.get(BASE_URL + warc_file, stream=True)
+    response.raise_for_status()
+
+    for record in ArchiveIterator(response.raw, arc2warc=True):
+        if record.rec_type == "response":
+            url = record.rec_headers.get_header("WARC-Target-URI")
+            content = (
+                record.content_stream()
+                .read()
+                .decode(errors="replace", encoding="utf-8")
+            )
+            records.append([url, content])
+            counter += 1
+
+        if n_records_to_download is not None and counter >= n_records_to_download:
+            break
+
+    return records
+
+
+class DownloadCommoncrawlSegments(DaskTransformComponent):
+    def transform(
+        self, df: dd.DataFrame, n_records_to_download: t.Optional[int] = None
+    ) -> dd.DataFrame:
+        """Downloads Commoncrawl segments based on a list of WARC paths.
+        Args:
+            df: A Dask DataFrame containing a column of WARC paths.
+            n_webpages_to_download: The number of webpages to download from each segment.
+        Returns:
+            A Dask DataFrame containing the downloaded webpages.
+        """
+        segment_paths = df["segment_path"].to_bag()
```

Comment:
What is the reason to use the `bag` interface here?

Reply:
So we can use dask `map` and `flatten` operations in the next steps. Using a list caused errors when converting the results to a dask dataframe later.

Reply:
Yes, but I mean compared to the dataframe API, where we have the `apply` function which behaves similarly.